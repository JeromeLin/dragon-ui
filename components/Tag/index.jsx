import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

class Tag extends Component {
  render() {
    const { props } = this;
    const {
      theme,
      size,
      isRadius,
      isRound,
      isActive,
      isFocus,
      isDisabled,
      className,
      onClose,
      children,
      style
    } = props;
    const disabled = 'disabled' in props || isDisabled;

    const cls = classnames({
      'ui-tag': true,
      radius: 'radius' in props || isRadius,
      round: 'round' in props || isRound,
      active: 'active' in props || isActive,
      focus: 'focus' in props || isFocus,
      disabled,
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
      [className]: !!className
    });

    const closeIcon = onClose ? (
      <Icon type="wrong" onClick={!disabled && onClose} />
    ) : null;
    return (
      <div className={cls} style={style}>
        {children}
        {closeIcon}
      </div>
    );
  }
}

Tag.propTypes = {
  theme: PropTypes.oneOf(['default', 'info', 'success', 'warning', 'error']),
  size: PropTypes.oneOf(['xl', 'lg', 'sm', 'xs']),
  onClose: PropTypes.func
};

Tag.defaultProps = {
  theme: 'default',
  size: null,
  onClose: () => {}
};

export default Tag;
