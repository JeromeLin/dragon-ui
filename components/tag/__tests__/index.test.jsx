import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tag from '../index';

describe('Tag', () => {
  it('renders normal Tag correctly', () => {
    const wrapper = render(
      <div className="multi-rows">
        <Tag>normal</Tag>
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Tag shape correctly', () => {
    const wrapper = render(
      <div>
        <Tag>radius</Tag>
        <Tag shape="rect">rect</Tag>
        <Tag shape="round">round</Tag>
      </div>,
    );
    expect(wrapper.find('.zw-tag--rect').length).toEqual(1);
    expect(wrapper.find('.zw-tag--round').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Tag color correctly', () => {
    const wrapper = render(
      <div className="multi-rows">
        <Tag color="red">normal</Tag>
      </div>,
    );
    expect(wrapper.find('.zw-tag--red').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Tag bordered correctly', () => {
    const unBorder = false;
    const wrapper = render(
      <div className="multi-rows">
        <Tag color="red">normal</Tag>
        <Tag color="green" bordered={unBorder}>unbordered</Tag>
      </div>,
    );
    expect(wrapper.find('.zw-tag--unborder').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Tag size correctly ', () => {
    const wrapper = render(
      <div className="multi-rows">
        <Tag>default</Tag>
        <Tag size="lg">lg</Tag>
        <Tag size="md">md</Tag>
        <Tag size="sm">sm</Tag>
        <Tag size="xs">xs</Tag>
      </div>,
    );
    expect(wrapper.find('.zw-tag').length).toEqual(5);
    expect(wrapper.find('.zw-tag--lg').length).toEqual(1);
    expect(wrapper.find('.zw-tag--sm').length).toEqual(1);
    expect(wrapper.find('.zw-tag--xs').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when closing tag', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Tag closable onClose={onClose}>可关闭标签</Tag>);
    expect(wrapper.find('.zw-icon').length).toEqual(1);
    wrapper.find('.zw-icon').simulate('click');
    expect(onClose).toHaveBeenCalled();
  });

  it('render CheckableTag correctly', () => {
    const checked = true;
    const wrapper = mount(<Tag.CheckableTag checked={checked}>default</Tag.CheckableTag>);

    expect(wrapper.find('.zw-tag--checked').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('render CheckableTag size correctly', () => {
    const wrapper = render(
      <div className="multi-rows">
        <Tag.CheckableTag size="lg">lg</Tag.CheckableTag>
        <Tag.CheckableTag size="md">md</Tag.CheckableTag>
        <Tag.CheckableTag size="sm">lg</Tag.CheckableTag>
        <Tag.CheckableTag size="xs">xs</Tag.CheckableTag>
      </div>,
    );
    expect(wrapper.find('.zw-tag').length).toEqual(4);
    expect(wrapper.find('.zw-tag--lg').length).toEqual(1);
    expect(wrapper.find('.zw-tag--sm').length).toEqual(1);
    expect(wrapper.find('.zw-tag--xs').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('behaves correctly when click CheckableTag', () => {
    const onChange = jest.fn();
    const checked = true;
    const wrapper = mount(<Tag.CheckableTag checked={checked} onChange={onChange}>default</Tag.CheckableTag>);

    expect(wrapper.find('.zw-tag--checked').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('.zw-tag--checked').simulate('click', !checked);
    expect(onChange).toHaveBeenCalled();
  });

  it('behaves correctly when disabled CheckableTag clicked uncalled onChange', () => {
    const onChange = jest.fn();
    const disabled = true;
    const wrapper = mount(<Tag.CheckableTag disabled={disabled} onChange={onChange}>default</Tag.CheckableTag>);

    expect(wrapper.find('.zw-tag--checked').length).toEqual(0);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('.zw-tag').simulate('click', !disabled);
    expect(onChange).not.toHaveBeenCalled();
  });
});
