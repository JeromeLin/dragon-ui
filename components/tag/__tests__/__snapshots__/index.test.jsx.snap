// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag behaves correctly when click CheckableTag 1`] = `
<CheckableTag
  checked={true}
  onChange={[MockFunction]}
  prefixCls="zw-tag"
>
  <div
    className="zw-tag zw-tag--checkable zw-tag--checked"
    onClick={[Function]}
  >
    default
  </div>
</CheckableTag>
`;

exports[`Tag behaves correctly when closing tag 1`] = `
<div
  class="zw-tag"
>
  可关闭标签
  <i
    class="zw-icon zw-icon--wrong zw-icon--default zw-icon--md"
    theme="default"
    type="wrong"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
    >
      <use
        href="#wrong"
      />
    </svg>
  </i>
</div>
`;

exports[`Tag behaves correctly when disabled CheckableTag clicked uncalled onChange 1`] = `
<CheckableTag
  disabled={true}
  onChange={[MockFunction]}
  prefixCls="zw-tag"
>
  <div
    className="zw-tag zw-tag--checkable zw-tag--disabled"
    onClick={[Function]}
  >
    default
  </div>
</CheckableTag>
`;

exports[`Tag render CheckableTag correctly 1`] = `
<CheckableTag
  checked={true}
  prefixCls="zw-tag"
>
  <div
    className="zw-tag zw-tag--checkable zw-tag--checked"
    onClick={[Function]}
  >
    default
  </div>
</CheckableTag>
`;

exports[`Tag render CheckableTag size correctly 1`] = `
<div
  class="multi-rows"
>
  <div
    class="zw-tag zw-tag--checkable zw-tag--lg"
  >
    lg
  </div>
  <div
    class="zw-tag zw-tag--checkable zw-tag--md"
  >
    md
  </div>
  <div
    class="zw-tag zw-tag--checkable zw-tag--sm"
  >
    lg
  </div>
  <div
    class="zw-tag zw-tag--checkable zw-tag--xs"
  >
    xs
  </div>
</div>
`;

exports[`Tag renders Tag bordered correctly 1`] = `
<div
  class="multi-rows"
>
  <div
    class="zw-tag zw-tag--red"
  >
    normal
  </div>
  <div
    class="zw-tag zw-tag--green zw-tag--unborder"
  >
    unbordered
  </div>
</div>
`;

exports[`Tag renders Tag color correctly 1`] = `
<div
  class="multi-rows"
>
  <div
    class="zw-tag zw-tag--red"
  >
    normal
  </div>
</div>
`;

exports[`Tag renders Tag shape correctly 1`] = `
<div>
  <div
    class="zw-tag"
  >
    radius
  </div>
  <div
    class="zw-tag zw-tag--rect"
  >
    rect
  </div>
  <div
    class="zw-tag zw-tag--round"
  >
    round
  </div>
</div>
`;

exports[`Tag renders Tag size correctly  1`] = `
<div
  class="multi-rows"
>
  <div
    class="zw-tag"
  >
    default
  </div>
  <div
    class="zw-tag zw-tag--lg"
  >
    lg
  </div>
  <div
    class="zw-tag zw-tag--md"
  >
    md
  </div>
  <div
    class="zw-tag zw-tag--sm"
  >
    sm
  </div>
  <div
    class="zw-tag zw-tag--xs"
  >
    xs
  </div>
</div>
`;

exports[`Tag renders normal Tag correctly 1`] = `
<div
  class="multi-rows"
>
  <div
    class="zw-tag"
  >
    normal
  </div>
</div>
`;
